courtlistener/process_cl.py

```diff
@@ -0,0 +1,132 @@
+import argparse
+import csv
+import logging
+import os
+import re
+import sys
+
+import pandas as pd
+
+from licensed_pile.licenses import PermissiveLicenses
+from licensed_pile.logs import configure_logging
+from licensed_pile.write import to_dolma
+
+SOURCE_NAME = "CourtListenerOpinion"
+
+csv.field_size_limit(sys.maxsize)
+
+logger = configure_logging("court-listener-opinion")
+
+
+def process_court_listener(file_path):
+    df = pd.read_csv(file_path)
+
+    # add metadata column
+    df["metadata"] = str(PermissiveLicenses.PD)
+
+    # add source column
+    df["source"] = SOURCE_NAME
+
+    """
+    Remove columns:
+    date_modified
+    author_str
+    per_curiam
+    joined_by_str
+    type
+    sha1
+    page_count
+    local_path
+    extracted_by_ocr
+    author_id
+    cluster_id
+    """
+    df = df.drop(
+        columns=[
+            "date_modified",
+            "author_str",
+            "per_curiam",
+            "joined_by_str",
+            "type",
+            "sha1",
+            "page_count",
+            "local_path",
+            "extracted_by_ocr",
+            "author_id",
+            "cluster_id",
+        ]
+    )
+
+    """
+    Merge columns based on Court Listener documentation:
+    html_with_citations
+    html_columbia
+    html_lawbox
+    xml_harvard
+    html_anon_2020
+    html
+    plain_text
+    """
+    df["text"] = (
+        df["html_with_citations"]
+        .combine_first(df["html_columbia"])
+        .combine_first(df["html_lawbox"])
+        .combine_first(df["xml_harvard"])
+        .combine_first(df["html_anon_2020"])
+        .combine_first(df["html"])
+    )
+
+    # keep only the text columns and drop null values
+    df = df.drop(
+        columns=[
+            "html",
+            "html_anon_2020",
+            "html_lawbox",
+            "html_columbia",
+            "xml_harvard",
+            "html_with_citations",
+        ]
+    ).dropna(subset=["text"])
+
+    # extract text from html and xml following Harvard CAP
+    # They used r"<.+?>", ""
+    df["text"] = df["text"].apply(lambda x: re.sub(r"<.+?>", "", x))
+
+    # combine merge plain text and extracted text
+    df["text"] = df["text"].combine_first(df["plain_text"])
+
+    # drop plain text column and text null values
+    df = df.drop(columns=["plain_text"]).dropna(subset=["text"])
+
+    # return a dictionary for each row - dolma format
+    return df.to_dict(orient="records")
```

Comment:
This doesn't result in the dolma formatted data. See https://github.com/allenai/dolma/blob/main/docs/data-format.md#dolma-document-format

Issues include:

* It has a `"download_url"` key top-level. This should be removed or moved into metadata as `"url"`
* It has `"date_created"` instead of `"created"`
* It is missing an `"added"` field
* It has the value of `"metadata"` as the license string. It should be a dict with a `"license"` key instead.

Reply:
I will create a metadata dict for the recommended changes.

Reply:
Let me know if this is what you were looking for:
```python
    df = pd.read_csv(file_path)

    df["source"] = args.source_name
    df["added"] = datetime.utcnow().isoformat()

    # rename the column date_created -> created
    df = df.rename(columns={"date_created": "created"})

    # add the metadata column including the author_str, license, and url
    df["metadata"] = df.apply(
        lambda x: {
            "license": str(PermissiveLicenses.PD),
            "url": x["download_url"],
            "author": x["author_str"],
        },
        axis=1,
    )
```